modules/resourceid/fsw/src/cfe_resourceid_api.c

```diff
@@ -0,0 +1,133 @@
+/*
+**  GSC-18128-1, "Core Flight Executive Version 6.7"
+**
+**  Copyright (c) 2006-2019 United States Government as represented by
+**  the Administrator of the National Aeronautics and Space Administration.
+**  All Rights Reserved.
+**
+**  Licensed under the Apache License, Version 2.0 (the "License");
+**  you may not use this file except in compliance with the License.
+**  You may obtain a copy of the License at
+**
+**    http://www.apache.org/licenses/LICENSE-2.0
+**
+**  Unless required by applicable law or agreed to in writing, software
+**  distributed under the License is distributed on an "AS IS" BASIS,
+**  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+**  See the License for the specific language governing permissions and
+**  limitations under the License.
+*/
+
+/*
+**  File:
+**  cfe_resource_api.c
+**
+**  Purpose:
+**      Function definitions related to CFE resource management
+**
+**  References:
+**     Flight Software Branch C Coding Standard Version 1.0a
+**     cFE Flight Software Application Developers Guide
+*/
+
+/*
+** Includes
+*/
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+
+#include "cfe.h"
+#include "cfe_resourceid.h"
+#include "cfe_resourceid_basevalue.h"
+
+/*********************************************************************/
+/*
+ * CFE_ResourceId_GetBase
+ *
+ * For complete API information, see prototype in header
+ */
+uint32 CFE_ResourceId_GetBase(CFE_ResourceId_t ResourceId)
+{
+    return (CFE_ResourceId_ToInteger(ResourceId) & ~((uint32)CFE_RESOURCEID_MAX));
+}
+
+/*********************************************************************/
+/*
+ * CFE_ResourceId_GetSerial
+ *
+ * For complete API information, see prototype in header
+ */
+uint32 CFE_ResourceId_GetSerial(CFE_ResourceId_t ResourceId)
+{
+    return (CFE_ResourceId_ToInteger(ResourceId) & ((uint32)CFE_RESOURCEID_MAX));
+}
+
+/*********************************************************************/
+/*
+ * CFE_ResourceId_ToIndex
+ *
+ * For complete API information, see prototype in header
+ */
+int32 CFE_ResourceId_ToIndex(CFE_ResourceId_t Id, uint32 BaseValue, uint32 TableSize, uint32 *Idx)
+{
+    uint32 Serial;
+
+    if (Idx == NULL)
+    {
+        return CFE_ES_BAD_ARGUMENT;
+    }
+
+    Serial = CFE_ResourceId_ToInteger(Id) - BaseValue;
```

Comment:
Should probably use  CFE_ResourceId_GetSerial(Id) to get the Serial here.

Reply:
The computation is intentionally different here - where ` CFE_ResourceId_GetSerial()` works with any type by simply masking out the bits, ignoring base - the intent here is to take the difference from the _expected_ base.

Even if we changed this to use `CFE_ResourceId_GetSerial()` the code would still need to check for a match to the expected base value.  Doing it this way just gets both checks done in one operation.